howso/boundary_values.amlg

```diff
@@ -0,0 +1,814 @@
+			;used for binary search
+			center (null)
+			left (null)
+			right (null)
+
+			;used for initial exponential search
+			query_delta (/ feature_residual 2)
+		)
+
+		;check lower side. start half a residual below and keep doubling the distance from the case value
+		;until the boundary condition is true (or go below min bound)
+		(while
+			(and
+				(> (- feature_value query_delta) (get feature_bounds "min"))
+				(not
+					(call !DoesNewValueCrossKnownBoundary (assoc new_feature_value (- feature_value query_delta) ))
+				)
+			)
+			(assign (assoc query_delta (* 2 query_delta) ))
+		)
+
+
+		;if the query_value is still greater than the min bound, then we must have crossed the boundary
+		(if (> (- feature_value query_delta) (get feature_bounds "min"))
+			(seq
+				;but now binary search between min bound and context value
+				(assign (assoc
+					left (- feature_value query_delta)
+					right (- feature_value (/ query_delta 2))
+				))
+				(assign (assoc
+					center
+						(if rounding_tuple
+							(apply "round" (append (/ (+ left right) 2) rounding_tuple))
+							(/ (+ left right) 2)
+						)
+				))
+
+				;binary search until within a residual
+				(while
+					(>
+						(- right center)
+						(get hyperparam_map ["featureResiduals" query_feature])
+					)
+
+					(if (call !DoesNewValueCrossKnownBoundary (assoc new_feature_value center))
+						;boundary still crossed, check right side
+						(assign (assoc left center ))
+
+						;else, boundary uncrossed, check left side
+						(assign (assoc right center ))
+					)
+					(assign (assoc
+						center
+							(if rounding_tuple
+								(apply "round" (append (/ (+ left right) 2) rounding_tuple))
+								(/ (+ left right) 2)
+							)
+					))
+				)
+
+				(assign (assoc lesser_boundary_value center))
+			)
+		)
+
+		;check upper side. start half a residual above and keep doubling the distance from the case value
+		;until the boundary condition is true (or go beyond max bound)
+		(assign (assoc query_delta (/ feature_residual 2) ))
+		(while
+			(and
+				(< (+ feature_value query_delta) (get feature_bounds "max"))
+				(not
+					(call !DoesNewValueCrossKnownBoundary (assoc new_feature_value (+ feature_value query_delta) ))
+				)
+			)
+			(assign (assoc query_delta (* 2 query_delta) ))
+		)
+
+
+		;if the query_value is still less than the max bound, then we must have crossed the boundary
+		(if (< (+ feature_value query_delta) (get feature_bounds "max"))
+			(seq
+				;but now binary search between min bound and context value
+				(assign (assoc
+					left (+ feature_value (/ query_delta 2))
+					right (+ feature_value query_delta)
+				))
+				(assign (assoc
+					center
+						(if rounding_tuple
+							(apply "round" (append (/ (+ left right) 2) rounding_tuple))
+							(/ (+ left right) 2)
+						)
+				))
+
+				;binary search until within a residual
+				(while
+					(>
+						(- right center)
+						; (get hyperparam_map ["featureResiduals" query_feature])
```

Comment:
should this be a TODO?

Reply:
err just missed flipping this back while testing